zebra/zebra_fpm.c

```diff
@@ -1276,6 +1472,219 @@ static int zfpm_trigger_update(struct route_node *rn, const char *reason)
+
+	return jhash_2words(mac_key, fpm_mac->vni, 0);
+}
+
+/*
+ * Compare function for FPM MAC info hash lookup
+ */
+static bool zfpm_mac_info_cmp(const void *p1, const void *p2)
+{
+	const struct fpm_mac_info_t *fpm_mac1 = p1;
+	const struct fpm_mac_info_t *fpm_mac2 = p2;
+
+	if (fpm_mac1 == NULL && fpm_mac2 == NULL)
+		return true;
+	if (fpm_mac1 == NULL || fpm_mac2 == NULL)
+		return false;
+
+	if (memcmp(fpm_mac1->macaddr.octet, fpm_mac2->macaddr.octet, ETH_ALEN)
+			!= 0)
+		return false;
+	if (memcmp(&fpm_mac1->r_vtep_ip, &fpm_mac2->r_vtep_ip,
+			sizeof(struct in_addr)) != 0)
+		return false;
+	if (fpm_mac1->vni != fpm_mac2->vni)
+		return false;
+
+	return true;
+}
+
+/*
+ * Lookup FPM MAC info hash entry.
+ */
+static struct fpm_mac_info_t *zfpm_mac_info_lookup(struct fpm_mac_info_t *key)
+{
+	if (!key)
+		return NULL;
+
+	return hash_lookup(zfpm_g->fpm_mac_info_table, key);
+}
+
+/*
+ * Callback to allocate fpm_mac_info_t structure.
+ */
+static void *zfpm_mac_info_alloc(void *p)
+{
+	const struct fpm_mac_info_t *key = p;
+	struct fpm_mac_info_t *fpm_mac;
+
+	fpm_mac = XCALLOC(MTYPE_FPM_MAC_INFO, sizeof(struct fpm_mac_info_t));
+	if (!fpm_mac)
+		return NULL;
+	memcpy(&fpm_mac->macaddr, &key->macaddr, ETH_ALEN);
+	memcpy(&fpm_mac->r_vtep_ip, &key->r_vtep_ip, sizeof(struct in_addr));
+	fpm_mac->vni = key->vni;
+
+	return ((void *)fpm_mac);
+}
+
+/*
+ * Delink and free fpm_mac_info_t.
+ */
+static void zfpm_mac_info_del(struct fpm_mac_info_t *fpm_mac)
+{
+	hash_release(zfpm_g->fpm_mac_info_table, fpm_mac);
+	TAILQ_REMOVE(&zfpm_g->mac_q, fpm_mac, fpm_mac_q_entries);
+	XFREE(MTYPE_FPM_MAC_INFO, fpm_mac);
+}
+
+/*
+ * zfpm_trigger_rmac_update
+ *
+ * Zebra code invokes this function to indicate that we should
+ * send an update to FPM for given MAC entry.
+ *
+ * This function checks if we already have enqueued an update for this RMAC,
+ * If yes, update the same fpm_mac_info_t. Else, create and enqueue an update.
+ */
+static int zfpm_trigger_rmac_update(zebra_mac_t *rmac, zebra_l3vni_t *zl3vni,
+					bool delete, const char *reason)
+{
+	char buf[ETHER_ADDR_STRLEN];
+	struct fpm_mac_info_t *fpm_mac, key;
+	struct interface *vxlan_if, *svi_if;
+
+	/*
+	 * Ignore if the connection is down. We will update the FPM about
+	 * all destinations once the connection comes up.
+	 */
+	if (!zfpm_conn_is_up())
+		return 0;
+
+	if (reason) {
+		zfpm_debug("triggering update to FPM - Reason: %s - %s",
+			reason,
+			prefix_mac2str(&rmac->macaddr, buf, sizeof(buf)));
+	}
+
+	vxlan_if = zl3vni_map_to_vxlan_if(zl3vni);
+	svi_if = zl3vni_map_to_svi_if(zl3vni);
+
+	memset(&key, 0, sizeof(struct fpm_mac_info_t));
+
+	memcpy(&key.macaddr, &rmac->macaddr, ETH_ALEN);
+	memcpy(&key.r_vtep_ip, &rmac->fwd_info.r_vtep_ip,
+		sizeof(struct in_addr));
+	key.vni = zl3vni->vni;
+
+	/* Check if this MAC is already present in the queue. */
+	fpm_mac = zfpm_mac_info_lookup(&key);
+
+	if (fpm_mac) {
+		if (!!CHECK_FLAG(fpm_mac->fpm_flags, ZEBRA_MAC_DELETE_FPM)
```

Comment:
we do not need a !!CHECK_FLAG.  None of the other places we use this macro do we use this.  I want consistency in style

Reply:
fpm_mac->fpm_flags is a uint32_t, delete is boolean. How can we do this if we don't use "!!CHECK_FLAG" ?

Reply:
All casts from integers to boolean are implicit, you don't need to invert to 1
```
vagrant@ubuntu-bionic ~> cat main.c
#include <stdio.h>
#include <stdint.h>
#include <inttypes.h>
#include <stdbool.h>

int main() {

        int test<:5:> = <%1, 2, 3, 4, 5%>;

        for (int i = 0; i < 5; i++) {
                bool b = i<:test:>;

                fprintf(stdout, "b as integer is: %d\n", b);
        }

        return 0;
}
vagrant@ubuntu-bionic ~> ./a.out
b as integer is: 1
b as integer is: 1
b as integer is: 1
b as integer is: 1
b as integer is: 1
```

Reply:
Okay, ignore both the above, that is different from what is being done here. @donaldsharp there is a comparison of int to bool; the int needs to be converted to bool somehow, hence the `!!`. 5 is true but `5 == true` is false.